### PR TITLE
Add caffe2_to_onnx converter and caffe2 models

### PR DESCRIPTION
**Pull request by jkamelin**

### CODE REVIEW

File: models/public/resnet-50-cf2/model.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+description: >-
+  This is an Caffe2\* version of `resnet-50` model, designed to perform image classification.
+  This model was converted from Caffe\* to Caffe2\* fromat. 
+  For details see repository <https://github.com/caffe2/models/tree/master/resnet50>,
+  paper <https://arxiv.org/pdf/1512.03385.pd>
```
Comment (zmaslova) [inline models/public/resnet-50-cf2/model.yml]: `.pdf`

File: tools/downloader/converter.py [hunk 2]
```diff
@@ -57,14 +57,21 @@ def prefixed_subprocess(prefix, args):
...
 
 def convert_to_onnx(model, output_dir, args, stdout_prefix):
-    pytorch_converter = Path(__file__).absolute().parent / 'pytorch_to_onnx.py'
+    if model.pytorch_to_onnx_args:
+        converter = Path(__file__).absolute().parent / 'pytorch_to_onnx.py'
+        converter_args = [string.Template(arg).substitute(conv_dir=output_dir / model.subdirectory,
+                                                            dl_dir=args.download_dir / model.subdirectory)
+                                for arg in model.pytorch_to_onnx_args]
+    if model.caffe2_to_onnx_args:
+        converter = Path(__file__).absolute().parent / 'caffe2_to_onnx.py'
```
Comment (IRDonch) [inline tools/downloader/converter.py]: Why not use the `caffe2.python.onnx.bin.conversion` module that's included in Caffe2?

File: models/public/densenet-121-cf2/densenet-121-cf2.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
+# densenet-121-cf2
+
+## Use Case and High-Level Description
+
+This is an Caffe2\* version of `densenet-121` model, one of the DenseNet
```
Comment (IRDonch) [inline models/public/densenet-121-cf2/densenet-121-cf2.md]: _a_ Caffe2 version

File: models/public/densenet-121-cf2/densenet-121-cf2.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+group of models designed to perform image classification. This model
+was converted from Caffe\* to Caffe2\* fromat.
```
Comment (IRDonch) [inline models/public/densenet-121-cf2/densenet-121-cf2.md]: format

File: models/public/densenet-121-cf2/densenet-121-cf2.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
+# densenet-121-cf2
+
+## Use Case and High-Level Description
+
+This is an Caffe2\* version of `densenet-121` model, one of the DenseNet
+group of models designed to perform image classification. This model
+was converted from Caffe\* to Caffe2\* fromat. 
+For details see repository <https://github.com/caffe2/models/tree/master/densenet121>,
+paper <https://arxiv.org/pdf/1608.06993.pdf>
```
Comment (IRDonch) [inline models/public/densenet-121-cf2/densenet-121-cf2.md]: Missing period at the end.

File: models/public/densenet-121-cf2/densenet-121-cf2.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+### Converted model
+
+Image, name - `data`,  shape - `1,3,224,224`, format is `B,C,H,W` where:
+
+- `B` - batch size
+- `C` - channel
+- `H` - height
+- `W` - width
+
+Channel order is `BGR`
```
Comment (IRDonch) [inline models/public/densenet-121-cf2/densenet-121-cf2.md]: Missing period at the end.

File: models/public/densenet-121-cf2/densenet-121-cf2.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
+# densenet-121-cf2
```
Comment (IRDonch) [inline models/public/densenet-121-cf2/densenet-121-cf2.md]: I think the suffix in the name should correspond to the identifier used in the `framework` field of the model field (i.e. `-caffe2`). Otherwise, we're just multiplying identifiers for no reason.

File: tools/downloader/requirements-caff2.in [hunk 8]
```diff
@@ -0,0 +1,3 @@
+future
```
Comment (IRDonch) [inline tools/downloader/requirements-caff2.in]: Missing `e` in the file name.

File: tools/downloader/README.md [hunk 9]
```diff
@@ -199,7 +200,7 @@ The basic usage is to run the script like this:
 ```
 
 This will convert all models into the Inference Engine IR format. Models that
-were originally in that format are ignored. Models in PyTorch's format will be
+were originally in that format are ignored. Models in PyTorch's and Caffe2 format will be
```
Comment (IRDonch) [inline tools/downloader/README.md]: Inconsistent grammar: "PyTorch's and Caffe2". Also, please don't add trailing whitespace.

File: models/public/densenet-121-cf2/densenet-121-cf2.md (hunk 7 above)
Comment (jkamelin) [reply models/public/densenet-121-cf2/densenet-121-cf2.md]: `-cf2` suffix is used for consistency with `-tf`

File: tools/accuracy_checker/configs/resnet-50-cf2.yml [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        - type: normalization
+          mean: 103.53, 116.28, 123.675
+          std: 57.375, 57.12, 58.395
+
+  - name: resnet-50-cf2
+    launchers:
+      - framework: dlsdk
+        tags:
+          - FP32
+        model: public/resnet-50-cf2/FP32/resnet-50-cf2.xml
```
Comment (eaidova) [inline tools/accuracy_checker/configs/resnet-50-cf2.yml]: please align model path start with weight path start for consistency with other AC configs. It is applicable for all added AC configs.

File: tools/downloader/README.md [hunk 11]
```diff
@@ -32,11 +32,12 @@ For the model converter, you will also need to install the OpenVINO&trade;
 toolkit and the prerequisite libraries for Model Optimizer. See the
 [OpenVINO toolkit documentation](https://docs.openvinotoolkit.org/) for details.
 
-If you using models from PyTorch framework, you will also need to use intermediate
+If you using models from PyTorch or Caffe2 framework, you will also need to use intermediate
 conversion to ONNX format. To use automatic conversion install additional dependencies:
 
 ```sh
 python3 -mpip install --user -r ./requirements-pytorch.in
+python3 -mpip install --user -r ./requirements-caffe2.in
```
Comment (eaidova) [inline tools/downloader/README.md]: does it need install both even user has model only in one framework format?

File: models/public/vgg19-cf2/model.yml [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+  - name: predict_net.pb
+    size: 2862
+    sha256: ebb8608fe80ee8bce096a60ecf3e6e8442eb118aaa0fa77d5d58b5dcff7dfb5f
+    source: https://s3.amazonaws.com/download.caffe2.ai/models/vgg19/predict_net.pb
+  - name: init_net.pb
+    size: 718338501
+    sha256: 492dbbbc7dd23cb052c66964714759f88370f3fa8542aa32556e93abc7beb69f
+    source: https://s3.amazonaws.com/download.caffe2.ai/models/vgg19/init_net.pb
+framework: caffe2
+caffe2_to_onnx:
```
Comment (eaidova) [inline models/public/vgg19-cf2/model.yml]: Wouldn't it be better to have unified field e.g. `convert_to_onnx` in the config and understand in the script what to convert based on the framework?
with the increase in the number of frameworks that require conversion to ONNX, we get an increase in script parameters that are unlikely to ever be used together (e.g. `pytorch_to_onnx_args`, `caffe2_to_onnx` args, `cntk_to_onnx_args` and so on)

File: models/public/densenet-121-cf2/densenet-121-cf2.md (hunk 7 above)
Comment (IRDonch) [reply models/public/densenet-121-cf2/densenet-121-cf2.md]: The difference with `tf` is that `tf` is also used as the `framework` field value, while `cf2` is not.

File: models/public/vgg19-cf2/model.yml (hunk 12 above)
Comment (IRDonch) [reply models/public/vgg19-cf2/model.yml]: +0, but it ought to be named something like `onnx_conversion_args`, for consistency with `model_optimizer_args`.

File: tools/downloader/README.md [hunk 13]
```diff
@@ -199,7 +204,7 @@ The basic usage is to run the script like this:
 ```
 
 This will convert all models into the Inference Engine IR format. Models that
-were originally in that format are ignored. Models in PyTorch's format will be
+were originally in that format are ignored. Models in PyTorch and Caffe2 format will be
```
Comment (IRDonch) [inline tools/downloader/README.md]: PyTorch and Caffe2 _formats_

File: tools/downloader/converter.py [hunk 14]
```diff
@@ -57,14 +57,24 @@ def prefixed_subprocess(prefix, args):
...
+    if model.pytorch_to_onnx_args:
+        converter = Path(__file__).absolute().parent / 'pytorch_to_onnx.py'
+        converter = [str(converter)]
+        converter_args = [string.Template(arg).substitute(conv_dir=output_dir / model.subdirectory,
+                                                            dl_dir=args.download_dir / model.subdirectory)
+                                for arg in model.pytorch_to_onnx_args]
+    if model.caffe2_to_onnx_args:
+        converter = ['-c', "import caffe2.python.onnx.bin.conversion;\
+                            import sys;\
+                            caffe2.python.onnx.bin.conversion.caffe2_to_onnx(sys.argv[1:])"]
```
Comment (IRDonch) [inline tools/downloader/converter.py]: You don't actually have to pass `sys.argv` to the entrypoint - if you pass no arguments, it will use `sys.argv` by default.

File: tools/downloader/caffe2_to_onnx.py [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+def convert_to_onnx(predict_net_path, init_net_path, input_shape, input_names, output_file, model_name=''):
+    """Convert Caffe2 model to ONNX and check the resulting onnx model"""
+
+    output_file.parent.mkdir(parents=True, exist_ok=True)
+
+    value_info = {}
+    data_type = onnx.TensorProto.FLOAT
+    input_names = input_names.split(',')
+    for name, shape in zip(input_names, input_shape):
+        value_info[name] = [data_type, shape]
```
Comment (IRDonch) [inline tools/downloader/caffe2_to_onnx.py]: There is a fair amount of code that parses `--input-shape` and `--input-names` and constructs `value_info` based on them. Why not just take `value_info` directly as an argument (like the `convert-caffe2-to-onnx` script does)?

File: tools/downloader/caffe2_to_onnx.py [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+    predict_net = caffe2_pb2.NetDef()
+    with open(predict_net_path, 'rb') as file:
```
Comment (IRDonch) [inline tools/downloader/caffe2_to_onnx.py]: It would be simpler to replace this with `predict_net_path.read_bytes()`. Likewise for other `open`s in this function.

File: tools/downloader/caffe2_to_onnx.py [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+
+def main():
+    args = parse_args()
+    convert_to_onnx(args.model_path, args.weights, args.input_shape,
+        args.input_names, args.output_file, args.model_name
+    )
+
+
+if __name__ == '__main__':
+    main()
```
Comment (IRDonch) [inline tools/downloader/caffe2_to_onnx.py]: Add a linebreak at the end, please.

File: tools/downloader/caffe2_to_onnx.py [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+    value_info = {}
+    data_type = onnx.TensorProto.FLOAT
+    input_names = input_names.split(',')
+    for name, shape in zip(input_names, input_shape):
+        value_info[name] = [data_type, shape]
+
+    predict_net = caffe2_pb2.NetDef()
+    with open(predict_net_path, 'rb') as file:
+        predict_net.ParseFromString(file.read())
+    if predict_net.name == "":
```
Comment (IRDonch) [inline tools/downloader/caffe2_to_onnx.py]: What is the purpose of this test? If a name was specified as a parameter, wouldn't we always want to use it?

File: tools/downloader/caffe2_to_onnx.py [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+    parser = argparse.ArgumentParser(description='Conversion of pretrained models from Caffe2 to ONNX')
+
+    parser.add_argument('--model-name', type=str, required=True,
+                        help='Model to convert. May be class name or name of constructor function')
+    parser.add_argument('--output-file', type=Path, required=True,
+                        help='Path to the output ONNX model')
+    parser.add_argument('--model-path', type=str, required=True,
+                        help='Path to predict_net .pb file')
+    parser.add_argument('--weights', type=str, required=True,
+                        help='Path to init_net .pb file')
```
Comment (IRDonch) [inline tools/downloader/caffe2_to_onnx.py]: The names of these parameters seem to be unnecessarily obscure. Why not call them `--predict-net` & `--init-net`, or something along those lines?

File: tools/downloader/caffe2_to_onnx.py [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+        result.append(single_shape)
+    return result
+
+def parse_args():
+    """Parse input arguments"""
+
+    parser = argparse.ArgumentParser(description='Conversion of pretrained models from Caffe2 to ONNX')
+
+    parser.add_argument('--model-name', type=str, required=True,
+                        help='Model to convert. May be class name or name of constructor function')
```
Comment (IRDonch) [inline tools/downloader/caffe2_to_onnx.py]: The help message for this is clearly wrong. Also, I have to ask: is it actually necessary to set the model name? What'll happen if we don't?

File: ci/requirements-conversion.txt [hunk 21]
```diff
@@ -2,8 +2,10 @@ absl-py==0.8.0            # via tensorboard, tensorflow
 astor==0.8.0              # via tensorflow
 certifi==2019.9.11        # via requests
 chardet==3.0.4            # via requests
+click==7.0
```
Comment (IRDonch) [inline ci/requirements-conversion.txt]: `click` shouldn't be required anymore (in fact, I see you've already removed it from the `.in` file).

File: tools/downloader/caffe2_to_onnx.py (hunk 19 above)
Comment (eizamaliev) [reply tools/downloader/caffe2_to_onnx.py]: It is done for consistent with `pytorch_to_onnx.py` parameter name.

Comment (IRDonch) [reply tools/downloader/caffe2_to_onnx.py]: But they don't have the same semantics as those parameters. Why should they be consistent with them?

File: tools/downloader/caffe2_to_onnx.py (hunk 15 above)
Comment (jkamelin) [reply tools/downloader/caffe2_to_onnx.py]: parameters `--input-shape`, `--input-names` is needed for consistency with `pytorch_to_onnx` converter, besides here we can check if the input shape values are nonnegative.

File: tools/downloader/caffe2_to_onnx.py (hunk 19 above)
Comment (jkamelin) [reply tools/downloader/caffe2_to_onnx.py]: Actually the semantics is quite the same, `init-net` stores all the weights, offsets, etc. used by the network, while the `predict-net` stores operations that that execute the net.

File: tools/downloader/caffe2_to_onnx.py (hunk 20 above)
Comment (jkamelin) [reply tools/downloader/caffe2_to_onnx.py]: The `onnx.checker` wont pass if the `net.name` is not set.

File: tools/downloader/caffe2_to_onnx.py (hunk 15 above)
Comment (IRDonch) [reply tools/downloader/caffe2_to_onnx.py]: Fair enough. However, your syntax for `--input-shape` doesn't match that used in `pytorch_to_onnx`. If you're going to claim consistency, you should align the syntaxes.

(This does imply that you'll only be able to support a single input shape, but it looks like that'll be enough for this PR anyway.)

File: tools/downloader/caffe2_to_onnx.py [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+
+def convert_to_onnx(predict_net_path, init_net_path, input_shape, input_names, output_file, model_name=''):
+    """Convert Caffe2 model to ONNX and check the resulting onnx model"""
+
+    output_file.parent.mkdir(parents=True, exist_ok=True)
+
+    value_info = {}
+    data_type = onnx.TensorProto.FLOAT
+    input_names = input_names.split(',')
+    for name, shape in zip(input_names, input_shape):
```
Comment (IRDonch) [inline tools/downloader/caffe2_to_onnx.py]: This can be simplified to `value_info = {name: [data_type, shape] for name, shape in zip(input_names, input_shape)}`.

File: tools/downloader/tests/representative-models.lst [hunk 23]
```diff
@@ -6,3 +6,4 @@ mtcnn-p # Caffe, HTTPS downloads, regex replacement
 octave-densenet-121-0.125 # MXNet, archive unpacking
 resnet-50-pytorch # PyTorch (torchvision)
 single-image-super-resolution-1032 # DLDT
+resnet-50-caffe2 # Caffe2, HTTPS downloads
```
Comment (IRDonch) [inline tools/downloader/tests/representative-models.lst]: In alphabetical order, please. Also, the comments are meant to list the features not covered by any preceding models, so please remove the "HTTPS downloads" bit.

**Comment by IRDonch** [comment]
@jkamelin Is this still WIP? Please update the title if not.

**Comment by IRDonch** [comment]
Unfortunately, `dl.caffe.berkeleyvision.org` seems to be down, which is blocking the downloader job. I'm going to give it some time to go back up, but if it doesn't, we might have to delete the three models we download from there. :-(

@vladimir-dudnik FYI

**Comment by IRDonch** [comment]
Jenkins please retry a build
